### ISSUE TITLE
systemd doesn't play nice with showmetrics

### ISSUE THREAD

**Issue by ageis** [zcash/zcash#1612]
Filing this as a known consequence of #1375. When running Zcash as a systemd service, you must specify `showmetrics=0`, otherwise it will fail with a floating point exception. I presume this is because systemd / journald doesn't like the ASCII art and special characters in the standard output.

```
root@home:~# systemctl status zcashd
● zcashd.service - Zcash daemon
   Loaded: loaded (/etc/systemd/system/zcashd.service; disabled; vendor preset: enabled)
   Active: failed (Result: start-limit-hit) since Sun 2016-10-23 02:02:57 PDT; 1s ago
  Process: 12022 ExecStart=/home/zcash/zcash/src/zcashd (code=killed, signal=FPE)
 Main PID: 12022 (code=killed, signal=FPE)

Oct 23 02:02:57 helios systemd[1]: zcashd.service: Main process exited, code=killed, status=8/FPE
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Unit entered failed state.
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Failed with result 'signal'.
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Service hold-off time over, scheduling restart.
Oct 23 02:02:57 helios systemd[1]: Stopped Zcash daemon.
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Start request repeated too quickly.
Oct 23 02:02:57 helios systemd[1]: Failed to start Zcash daemon.
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Unit entered failed state.
Oct 23 02:02:57 helios systemd[1]: zcashd.service: Failed with result 'start-limit-hit'.
```

Example Zcash systemd service unit:
```
[Unit]
Description=Zcash daemon
After=network.target

[Service]
Type=simple
User=zcash
WorkingDirectory=/home/zcash/.zcash
ExecStart=/home/zcash/zcash/src/zcashd
Restart=always

[Install]
WantedBy=multi-user.target
```

**Comment by daira** [comment]
It might not be anything to do with the output itself; there is a floating point division to calculate the solution rate (and I think we removed all other use of floating point).

**Comment by daira** [comment]
We could easily replace that division with an integer (fixed-point) one. Try commenting out these three lines for now: https://github.com/zcash/zcash/blob/master/src/metrics.cpp#L115-L117 and see whether that fixes it.

**Comment by str4d** [comment]
Also, this should not be a problem anyway, as for `zcashd -daemon` the metrics screen is automatically disabled. Is there another way we should be detecting that `zcashd` is not being run in a Bash console?

**Comment by daira** [comment]
I'd prefer to avoid fp entirely if there are some situations where it breaks. Yes, it is possible to detect an interactive terminal: http://stackoverflow.com/a/7601564/393146

**Comment by arithmetric** [comment]
I've reproduced this error in a few environments where a TTY is not allocated: when zcashd is started as a service by Supervisor, when started in a Docker container without a TTY (docker run without `-t`), and when output is piped to a file (simply `zcashd > zcash.log`).

I haven't tested this yet, but I suspect that in non-TTY environments, the call to ioctl() to get window size is not returning a positive integer for number of columns.

See: https://github.com/zcash/zcash/blob/master/src/metrics.cpp#L198

The number of columns is then used as a divisor here:
https://github.com/zcash/zcash/blob/master/src/metrics.cpp#L110
https://github.com/zcash/zcash/blob/master/src/metrics.cpp#L145

If this is causing the error, could a sensible default (like 80) be used for the number of columns?

**Comment by arithmetric** [comment]
I was able to confirm a fix for this by providing a default number of columns and only using the value from ioctl() when it is non-zero (and only in TTY environments).

I've submitted #1615 with my approach to fix this.

**Comment by str4d** [comment]
This was closed by #1615.
